### PR TITLE
BUG: Avoid accessing invalid page when inserting blank page under some conditions

### PR DESCRIPTION
**Pull request by j-t-1**
width and height are now treated symmetrically and updated only if they are not provided.

### CODE REVIEW

**Review by stefan6419846 — CHANGES_REQUESTED** [review]
* Please add a test for an out-of-bounds index.
* What happens if the width is 0 or negative?
* Please update the title to reflect what the actual bug was - currently, it is not obvious which logic is meant or what was wrong.

**Comment by j-t-1** [comment]
> What happens if the width is 0 or negative?
> Is the code okay?

`PageObject.create_blank_page(self, width, height)` gets created with 0 or negative values. Same for height.

If `width` or `height` is 0 or negative change use old_page values?

**Comment by stefan6419846** [comment]
The specific case from https://github.com/py-pdf/pypdf/pull/3337#discussion_r2541790250 is still not tested?

**Comment by j-t-1** [comment]
Shall I do this?
`if (width is None or width <= 0 or height is None or height <= 0) and index < self.get_num_pages():`

**Comment by j-t-1** [comment]
> The specific case from [#3337 (comment)](https://github.com/py-pdf/pypdf/pull/3337#discussion_r2541790250) is still not tested?

[Add a test for an out-of-bounds index](https://github.com/py-pdf/pypdf/pull/3529/commits/476942bbb9ed8b2012e6ca9748a05a6b30c1e74e) does this I thought.

Have I not done the test correctly?

**Comment by stefan6419846** [comment]
If would use `if not (width or height) and index <= self.get_num_pages()` should be sufficient - negative indices are a common edge case I would not look into and just allow negative list indexing.

> [Add a test for an out-of-bounds index](https://github.com/py-pdf/pypdf/pull/3529/commits/476942bbb9ed8b2012e6ca9748a05a6b30c1e74e) does this I thought. Have I not done the test correctly?

Have you tested the behavior without your change to the logic? Did it fail? If not, your change most likely does not yet cover the desired case.

**Comment by j-t-1** [comment]
> Have you tested the behavior without your change to the logic? Did it fail? If not, your change most likely does not yet cover the desired case.

No. So I do the same check on the old code and make sure it fails?

**Comment by stefan6419846** [comment]
Yes. This is the usual way you should assert that your tests are correct.

**Comment by j-t-1** [comment]
> If would use `if not (width or height) and index <= self.get_num_pages()` should be sufficient - negative indices are a common edge case I would not look into and just allow negative list indexing.

Thanks.

> Have you tested the behavior without your change to the logic? Did it fail? If not, your change most likely does not yet cover the desired case.

Will do. Is this the only outstanding now?

**Comment by stefan6419846** [comment]
> Will do. Is this the only outstanding now?

As far as I can currently see, yes. We probably want to cover both variants (setting only width, setting only height), although only one case has previously been broken, but it makes the tests more consistent.

**Comment by j-t-1** [comment]
The old code:

```
        if width is None or (height is None and index < self.get_num_pages()):
            oldpage = self.pages[index]
            width = oldpage.mediabox.width
            height = oldpage.mediabox.height
```
Has this error:

```
        old_page = writer.pages[0]
        page = writer.insert_blank_page(width=10, index=0)
>       assert page.mediabox.width == 10
E       AssertionError: assert 612.0 == 10
E        +  where 612.0 = RectangleObject([0.0, 0.0, 612, 792]).width
E        +    where RectangleObject([0.0, 0.0, 612, 792]) = {'/Type': '/Page', '/Parent': NullObject, '/Resources': {}, '/MediaBox': RectangleObject([0.0, 0.0, 612, 792])}.mediabox
```

Because the `height` is None, and thus the `width` does not get updated. Whereas this PR passes this test.

**Comment by stefan6419846** [comment]
But we still do not have a test for the previous wrong code mentioned before?

```python
page = writer.insert_blank_page(width=None, height=42, index=len(writer.pages))
```

**Comment by j-t-1** [comment]
> But we still do not have a test for the previous wrong code mentioned before?
> 
> ```python
> page = writer.insert_blank_page(width=None, height=42, index=len(writer.pages))
> ```


Good call. I think the code is incorrect, as it fails this test.

File: tests/test_writer.py [hunk 1]
```diff
@@ -255,6 +259,57 @@ def writer_operate(writer: PdfWriter) -> None:
...
+    assert page.mediabox.height == old_page.mediabox.height
+
+    old_page_count = len(writer.pages)
+    old_page = writer.pages[old_page_count - 1]
+    page = writer.insert_blank_page(width=50, height=60, index=len(writer.pages))
+    assert len(writer.pages) == old_page_count + 1
+    assert page.mediabox.width == 50
+    assert page.mediabox.height == 60
+
+    old_page_count = len(writer.pages)
```
Comment (stefan6419846) [inline tests/test_writer.py]: This variable is unused.

Comment (j-t-1) [reply tests/test_writer.py]: Amended.

File: pypdf/_writer.py [hunk 2]
```diff
@@ -713,10 +711,15 @@ def insert_blank_page(
                 and previous page does not exist.
 
         """
-        if width is None or (height is None and index < self.get_num_pages()):
-            oldpage = self.pages[index]
-            width = oldpage.mediabox.width
-            height = oldpage.mediabox.height
+        if (width is None or width <= 0) and index < self.get_num_pages():
```
Comment (stefan6419846) [inline pypdf/_writer.py]: Accessing `self.get_num_pages()` three times without changing the page count in-between is costly. Consider saving it to a variable once and re-use it for the different checks.

File: pypdf/_writer.py [hunk 3]
```diff
@@ -713,10 +711,15 @@ def insert_blank_page(
...
+            width = self.pages[index].mediabox.width
+        if (height is None or height <= 0) and index < self.get_num_pages():
+            height = self.pages[index].mediabox.height
```
Comment (stefan6419846) [inline pypdf/_writer.py]: If both width and height are `None`, this requires accessing the same page twice. Are we able to refactor the condition?

Example: Check the index first and assign the page, otherwise set the page to `None`. Then rewrite the two existing conditions to check for `page is not None`.

File: pypdf/_writer.py [hunk 4]
```diff
@@ -713,10 +711,15 @@ def insert_blank_page(
...
+
+        if index == self.get_num_pages():
+            width = self.pages[index - 1].mediabox.width if width is None or width <= 0 else width
```
Comment (stefan6419846) [inline pypdf/_writer.py]: `self.pages[index - 1].mediabox` could be saved for further use.

**Review by stefan6419846 — CHANGES_REQUESTED** [review]

File: pypdf/_writer.py (hunk 4 above)
Comment (j-t-1) [reply pypdf/_writer.py]: Done.

File: pypdf/_writer.py (hunk 2 above)
Comment (j-t-1) [reply pypdf/_writer.py]: Done.

File: pypdf/_writer.py (hunk 3 above)
Comment (j-t-1) [reply pypdf/_writer.py]: I have refactored this so the code blocks match. Is this okay?

File: pypdf/_writer.py [hunk 5]
```diff
@@ -713,10 +711,19 @@ def insert_blank_page(
...
 
         """
-        if width is None or (height is None and index < self.get_num_pages()):
-            oldpage = self.pages[index]
-            width = oldpage.mediabox.width
-            height = oldpage.mediabox.height
+        num_pages = self.get_num_pages()
+        if index < num_pages:
+            mediabox = self.pages[index].mediabox
+            if (width is None or width <= 0):
```
Comment (stefan6419846) [inline pypdf/_writer.py]: The parentheses are not needed any more.

File: pypdf/_writer.py [hunk 6]
```diff
@@ -713,10 +711,19 @@ def insert_blank_page(
...
+                width = mediabox.width
+            if (height is None or height <= 0):
+                height = mediabox.height
+
+        if index == num_pages:
```
Comment (stefan6419846) [inline pypdf/_writer.py]: It seems like we can still simplify this and not have too much duplicates here. Both new conditions are basically the same, except that for `index == num_pages`, we choose the last.

I guess this could be written like this?

```python
num_pages = self.get_num_pages()
if index <= num_pages:
    # Use the chosen index, but do not exceed the available pages.
    fixed_index = min(index, num_pages - 1)
    mediabox = self.pages[fixed_index].mediabox
    if width is None or width <= 0:
        width = mediabox.width
    if height is None or height <= 0:
        height = mediabox.height
```

Comment (j-t-1) [reply pypdf/_writer.py]: Nice.

File: pypdf/_writer.py (hunk 5 above)
Comment (j-t-1) [reply pypdf/_writer.py]: Removed.

**Comment by j-t-1** [comment]
What do you want done `if index > num_pages`?

**Comment by stefan6419846** [comment]
We could theoretically remove the condition completely and would use the last page in this case (which should be documented and tested) - or raise an error as this is a user error. I would probably go with a user error, as this should never be what the user wants.

**Comment by j-t-1** [comment]
> We could theoretically remove the condition completely and would use the last page in this case (which should be documented and tested) - or raise an error as this is a user error. I would probably go with a user error, as this should never be what the user wants.

Good call. I will amend so it raises an error.

**Comment by j-t-1** [comment]
@stefan6419846 this got stuck in the tests, please could you re-run it if you have that option?

**Comment by stefan6419846** [comment]
Something is odd with the PyPy 3.11 tests lately, as they regularly get stuck. I have not yet found a pattern except re-triggering it (possibly after a rebase) and hoping.

File: pypdf/_writer.py [hunk 7]
```diff
@@ -711,12 +709,20 @@ def insert_blank_page(
         Raises:
             PageSizeNotDefinedError: if width and height are not defined
                 and previous page does not exist.
+            IndexError: Index is outside of [0, self.get_num_pages()]
+        """
+        num_pages = self.get_num_pages()
+        if index <= num_pages:
```
Comment (stefan6419846) [inline pypdf/_writer.py]: As we allow negative indexing, we should probably use the absolute value for the check to avoid "underflows" and add tests accordingly.

```suggestion
        if abs(index) <= num_pages:
```

File: pypdf/_writer.py [hunk 8]
```diff
@@ -711,12 +709,20 @@ def insert_blank_page(
...
+            # Use the chosen index, but do not exceed the available pages
+            fixed_index = min(index, num_pages - 1)
+            mediabox = self.pages[fixed_index].mediabox
+            if width is None or width <= 0:
+                width = mediabox.width
+            if height is None or height <= 0:
+                height = mediabox.height
+        else:
+            raise IndexError("Index should be between zero and the number of existing pages")
```
Comment (stefan6419846) [inline pypdf/_writer.py]: This is misleading, as it only covers the range `[1, num_pages]`. We should at least have `[0, num_pages]` or even `[-num_pages, num_pages]`.

**Review by stefan6419846 — CHANGES_REQUESTED** [review]

File: pypdf/_writer.py (hunk 7 above)
Comment (j-t-1) [reply pypdf/_writer.py]: Negative indexing is where -1 is the last page, like in Python indexing?

Comment (stefan6419846) [reply pypdf/_writer.py]: As we are doing regular list indexing here, yes.

File: pypdf/_writer.py [hunk 9]
```diff
@@ -711,12 +709,20 @@ def insert_blank_page(
...
+        if abs(index) <= num_pages:
+            # Use the chosen index, but do not exceed the available pages
+            fixed_index = min(index, num_pages - 1)
+            mediabox = self.pages[fixed_index].mediabox
+            if width is None or width <= 0:
+                width = mediabox.width
+            if height is None or height <= 0:
+                height = mediabox.height
+        else:
+            raise IndexError("Index should be in [-num_pages, num_pages]")
```
Comment (stefan6419846) [inline pypdf/_writer.py]:
```suggestion
            raise IndexError(f"Index should be in range [-{num_pages}, {num_pages}]")
```

File: tests/test_writer.py [hunk 10]
```diff
@@ -251,6 +255,83 @@ def writer_operate(writer: PdfWriter) -> None:
...
+    writer = PdfWriter(clone_from=RESOURCE_ROOT / "crazyones.pdf")
+
+    old_page_count = len(writer.pages)
+    old_page = writer.pages[0]
+    page = writer.insert_blank_page(index=0)
+    assert len(writer.pages) == old_page_count + 1
+    assert page.mediabox.width == old_page.mediabox.width
+    assert page.mediabox.height == old_page.mediabox.height
+
+    old_page_count = len(writer.pages)
```
Comment (stefan6419846) [inline tests/test_writer.py]: Why do we have to determine this on each test? Couldn't we use the base value once and basically add the desired offset for each further check?

**Review by stefan6419846 — CHANGES_REQUESTED** [review]
Apart from the inline feedback: Shouldn't we test at least once valid negative index and one valid positive index `0 < index < num_pages` as well?

**Comment by j-t-1** [comment]
> Apart from the inline feedback: Shouldn't we test at least once valid negative index and one valid positive index `0 < index < num_pages` as well?

Done.

**Review by stefan6419846 — APPROVED** [review]
